### PR TITLE
add vulnerability scanning periodic job

### PR DESCRIPTION
**Pull request by sdowell**
This job checks the container analysis results in GCP for vulnerabilities on the postsubmit artifacts for the checked out commit.

### CODE REVIEW

**Comment by sdowell** [comment]
This job depends on https://github.com/GoogleContainerTools/kpt-config-sync/pull/607

File: prow/prowjobs/GoogleContainerTools/kpt-config-sync/kpt-config-sync-periodics.yaml [hunk 1]
```diff
@@ -1342,3 +1342,31 @@ periodics:
...
+  spec:
+    serviceAccountName: vulnerability-scanner
+    containers:
+    - image: us-docker.pkg.dev/kpt-config-sync-ci-artifacts/test-infra/vulnerability-scanner:v1.0.0-60ad07344
+      command:
+      - make
+      - vulnerability-scan-postsubmit
+      resources:
+        requests:
+          memory: "2Gi"
```
Comment (karlkfi) [inline prow/prowjobs/GoogleContainerTools/kpt-config-sync/kpt-config-sync-periodics.yaml:1368]: does it really need 2Gi?

Is it scanning locally or just calling gcloud for cloud vuln reports?

File: prow/prowjobs/GoogleContainerTools/kpt-config-sync/kpt-config-sync-periodics.yaml [hunk 2]
```diff
@@ -1342,3 +1342,31 @@ periodics:
...
+          cpu: "2000m"
```
Comment (karlkfi) [inline prow/prowjobs/GoogleContainerTools/kpt-config-sync/kpt-config-sync-periodics.yaml:1369]: Does this need the equivalent of 2 full cores?

File: prow/prowjobs/GoogleContainerTools/kpt-config-sync/kpt-config-sync-periodics.yaml [hunk 3]
```diff
@@ -1342,3 +1342,31 @@ periodics:
...
+
+- name: vulnerability-scan
+  interval: 1h
+  annotations:
+    testgrid-dashboards: googleoss-kpt-config-sync-misc
+    testgrid-tab-name: vulnerability-scan
+  cluster: build-kpt-config-sync
+  decorate: true
+  decoration_config:
+    timeout: 1h
```
Comment (karlkfi) [inline prow/prowjobs/GoogleContainerTools/kpt-config-sync/kpt-config-sync-periodics.yaml:1354]: will it ever take an hour or is this a conservative timeout?

File: prow/prowjobs/GoogleContainerTools/kpt-config-sync/kpt-config-sync-periodics.yaml [hunk 4]
```diff
@@ -1342,3 +1342,31 @@ periodics:
...
+    - image: us-docker.pkg.dev/kpt-config-sync-ci-artifacts/test-infra/vulnerability-scanner:v1.0.0-60ad07344
+      command:
+      - make
+      - vulnerability-scan-postsubmit
+      resources:
+        requests:
+          memory: "2Gi"
+          cpu: "2000m"
+    nodeSelector:
+      # This job requires 8vCPUs or less, so it is "small".
```
Comment (karlkfi) [inline prow/prowjobs/GoogleContainerTools/kpt-config-sync/kpt-config-sync-periodics.yaml:1371]: Why 8vCPUs if it only requests 2?

File: prow/prowjobs/GoogleContainerTools/kpt-config-sync/kpt-config-sync-periodics.yaml (hunk 1 above)
Comment (sdowell) [reply prow/prowjobs/GoogleContainerTools/kpt-config-sync/kpt-config-sync-periodics.yaml:1368]: No it probably doesn't need this much memory, this was admittedly just copy/pasted from a similar job. It just calls to gcloud so it shouldn't need too much memory, I can shrink this

File: prow/prowjobs/GoogleContainerTools/kpt-config-sync/kpt-config-sync-periodics.yaml (hunk 2 above)
Comment (sdowell) [reply prow/prowjobs/GoogleContainerTools/kpt-config-sync/kpt-config-sync-periodics.yaml:1369]: Probably not

File: prow/prowjobs/GoogleContainerTools/kpt-config-sync/kpt-config-sync-periodics.yaml (hunk 4 above)
Comment (sdowell) [reply prow/prowjobs/GoogleContainerTools/kpt-config-sync/kpt-config-sync-periodics.yaml:1371]: This comment came from copy paste but I think it's still valid, `8vCPUs` refers to the capacity of the machine type for this node pool

File: prow/prowjobs/GoogleContainerTools/kpt-config-sync/kpt-config-sync-periodics.yaml (hunk 3 above)
Comment (sdowell) [reply prow/prowjobs/GoogleContainerTools/kpt-config-sync/kpt-config-sync-periodics.yaml:1354]: It's a conservative timeout, it shouldn't take this long

File: prow/prowjobs/GoogleContainerTools/kpt-config-sync/kpt-config-sync-periodics.yaml (hunk 1 above)
Comment (sdowell) [reply prow/prowjobs/GoogleContainerTools/kpt-config-sync/kpt-config-sync-periodics.yaml]: reduced the requests

File: prow/prowjobs/GoogleContainerTools/kpt-config-sync/kpt-config-sync-periodics.yaml [hunk 5]
```diff
@@ -1342,3 +1342,31 @@ periodics:
...
+    testgrid-dashboards: googleoss-kpt-config-sync-misc
+    testgrid-tab-name: vulnerability-scan
+  cluster: build-kpt-config-sync
+  decorate: true
+  decoration_config:
+    timeout: 1h
+  extra_refs:
+  - org: GoogleContainerTools
+    repo: kpt-config-sync
+    base_ref: main
```
Comment (nan-yu) [inline prow/prowjobs/GoogleContainerTools/kpt-config-sync/kpt-config-sync-periodics.yaml:1358]: Just to confirm the vulnerability postsubmit job only runs against `main`?

Comment (sdowell) [reply prow/prowjobs/GoogleContainerTools/kpt-config-sync/kpt-config-sync-periodics.yaml:1358]: Yes this job currently only runs on `main`, but the intention is to add another job definition for the release branch once:
- the postsubmit changes are on the release branch (e.g. cherrypick) and the postsubmit job is running on the release branch
- the vulnerability scanning changes are on the release branch (e.g. cherrypick)

File: prow/prowjobs/GoogleContainerTools/kpt-config-sync/kpt-config-sync-periodics.yaml (hunk 3 above)
Comment (sdowell) [reply prow/prowjobs/GoogleContainerTools/kpt-config-sync/kpt-config-sync-periodics.yaml]: reduced to 30m

File: prow/prowjobs/GoogleContainerTools/kpt-config-sync/kpt-config-sync-periodics.yaml (hunk 5 above)
Comment (nan-yu) [reply prow/prowjobs/GoogleContainerTools/kpt-config-sync/kpt-config-sync-periodics.yaml:1358]: I'm wondering whether we need a job for the release branch. The release branch is created based off the main branch, and all commits are cherry-picked from main. If the main branch has no vulnerabilities, the release branch should have none, too. If the main branch has vulnerabilities, the release branch may or may not have, but eventually, the vulns should be fixed on main for the next minor release.

Comment (sdowell) [reply prow/prowjobs/GoogleContainerTools/kpt-config-sync/kpt-config-sync-periodics.yaml:1358]: I think we should still have a job for the release branch, as even if we are cherry picking changes to the release branch it's not guaranteed that all artifacts/dependencies will be identical between refs. We should verify artifacts from the exact commit SHA to be certain

Comment (nan-yu) [reply prow/prowjobs/GoogleContainerTools/kpt-config-sync/kpt-config-sync-periodics.yaml:1358]: Make sense

**Review by nan-yu — APPROVED** [review]
/lgtm

**Pull request merged**